### ISSUE TITLE
WordVectorSerializer.loadTxtVectors causes memory error with GPU backend

### ISSUE THREAD

**Issue by kottmann** [deeplearning4j/nd4j#1842]
#### Issue Description

This causes an ouf of memory error with  the CUDA 8.0 GPU backend but runs fine with the CPU backend. The  WordVectorSerializer.loadTxtVectors  was called to load the text versoin of glove.840B.300d.txt. That er roughly 2m vectors with a dimension of 300.

The system running this has 32 GB of memory, and 5g where allocated to the JVM via -Xmx. I tried with different settings and got the same error.
Here is the log output and error:
 [main] INFO org.nd4j.linalg.factory.Nd4jBackend - Loaded [JCublasBackend] backend
[main] INFO org.nd4j.nativeblas.NativeOpsHolder - Number of threads used for NativeOps: 32
[main] INFO org.reflections.Reflections - Reflections took 89 ms to scan 13 urls, producing 29 keys and 189 values 
[main] INFO org.nd4j.linalg.api.ops.executioner.DefaultOpExecutioner - Backend used: [CUDA]; OS: [Linux]
[main] INFO org.nd4j.linalg.api.ops.executioner.DefaultOpExecutioner - Cores: [8]; Memory: [4.4GB];
[main] INFO org.nd4j.linalg.api.ops.executioner.DefaultOpExecutioner - Blas vendor: [CUBLAS]
[main] INFO org.nd4j.linalg.jcublas.ops.executioner.CudaExecutioner - Device name: [GeForce GTX 1070]; CC: [6.1]; Total/free memory: [8504868864]
Exception in thread "main" java.lang.OutOfMemoryError: Cannot allocate new FloatPointer(300): totalBytes = 1, physicalBytes = 8G
	at org.bytedeco.javacpp.FloatPointer.<init>(FloatPointer.java:76)
	at org.bytedeco.javacpp.FloatPointer.<init>(FloatPointer.java:41)
	at org.nd4j.linalg.jcublas.buffer.BaseCudaDataBuffer.set(BaseCudaDataBuffer.java:349)
	at org.nd4j.linalg.jcublas.buffer.BaseCudaDataBuffer.<init>(BaseCudaDataBuffer.java:236)
	at org.nd4j.linalg.jcublas.buffer.CudaFloatDataBuffer.<init>(CudaFloatDataBuffer.java:84)
	at org.nd4j.linalg.jcublas.buffer.factory.CudaDataBufferFactory.createFloat(CudaDataBufferFactory.java:301)
	at org.nd4j.linalg.factory.Nd4j.createBuffer(Nd4j.java:1322)
	at org.nd4j.linalg.factory.BaseNDArrayFactory.create(BaseNDArrayFactory.java:1956)
	at org.nd4j.linalg.factory.Nd4j.create(Nd4j.java:3216)
	at org.nd4j.linalg.factory.Nd4j.create(Nd4j.java:3039)
	at org.deeplearning4j.models.embeddings.loader.WordVectorSerializer.loadTxt(WordVectorSerializer.java:1804)
	at org.deeplearning4j.models.embeddings.loader.WordVectorSerializer.loadTxtVectors(WordVectorSerializer.java:1714)
	at opennlp.tools.dl.NameFinderDL.main(NameFinderDL.java:205)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.intellij.rt.execution.application.AppMain.main(AppMain.java:147)
Caused by: java.lang.OutOfMemoryError: Physical memory usage is too high: physicalBytes = 8G > maxPhysicalBytes = 8G
	at org.bytedeco.javacpp.Pointer.deallocator(Pointer.java:572)
	at org.bytedeco.javacpp.Pointer.init(Pointer.java:121)
	at org.bytedeco.javacpp.FloatPointer.allocateArray(Native Method)
	at org.bytedeco.javacpp.FloatPointer.<init>(FloatPointer.java:68)
	... 17 more

#### Version Information

Please indicate relevant versions, including, if relevant:

* libnd4j version 0.8.1-SNAPSHOT / 9092509
* Nd4j version 0.8.1-SNAPSHOT / 71f99d1
* platform information Linux-x86_64 / Fedora 25
* CUDA version, if used 8.0
* NVIDIA driver version, if in use 381.22

**Comment by saudet** [comment]
`physicalBytes = 8G` -> You're giving the JVM less memory to work with than what you have as video memory?

**Comment by kottmann** [comment]
@saudet I tried multiple settings for the JVM heap, always got the out of memory exception, e.g. with 16g or 24g. Any suggestions on which value should work?

**Comment by saudet** [comment]
Your process used up to 24g of memory?

**Comment by kottmann** [comment]
Let me try again. What I can do to debug this? With the native backend it loads the vectors from the file within a minute, with GPU it crashes after something like an hour on my my system.

**Comment by agibsonccc** [comment]
@kottmann got some code we can look at? I believe we asked for that.
Please just use heap space, not the javacpp knob.
What GPU are you using? What's happening here is nd4j is loading the word vectors in to GPU ram (we double allocate cpu and gpu ram) - and your gpu is running out of memory thus it freezes.

**Comment by kottmann** [comment]
@agibsonccc take a look here https://github.com/apache/opennlp-sandbox/blob/master/opennlp-dl/src/main/java/opennlp/tools/dl/NameFinderDL.java#L205

We are trying to build a prototype to perform NER for Apache OpenNLP. I am using a GTX 1070 with 8GB.

Let me verify that and then report back here. It probably just doesn't make much sense to load that data the way I do here.

**Comment by agibsonccc** [comment]
Thanks - this is the discussion we had been talking about before.  I will /cc @raver119 who might have some ideas for some work arounds here as well. @kottmann we maintain the whole stack here - so there's some knobs we can likely pull for you.

**Comment by raver119** [comment]
WordVectorSerializer.loadStaticModel() was first attempt to solve this issue. In this case device memory won't be used for vectors during initialization.

However, now i think it will be better if we'll add some hybrid execution here, to simply avoid gpu use during initialization.
